pkg/ipcache/cidr.go

```diff
@@ -151,17 +144,28 @@ func (ipc *IPCache) UpsertGeneratedIdentities(newlyAllocatedIdentities map[netip
 		if !ok {
 			continue
 		}
-		if _, ok := ipc.LookupByIPRLocked(prefix.String()); ok {
-			// Already there; continue
+		existing, ok := ipc.LookupByIPRLocked(prefix.String())
+		if !ok {
+			// We need this identity, but it was somehow deleted
+			metrics.IPCacheErrorsTotal.WithLabelValues(
+				metricTypeRecover, metricErrorUnexpected,
+			).Inc()
+			toUpsert[prefix] = id
 			continue
 		}
-		toUpsert[prefix] = id
+		if existing.createdFromMetadata {
+			// the createdFromMetadata field is used to tell the ipcache that it is safe to delete
+			// a prefix when all entries are removed from the metadata layer. However, as this is the
+			// "old-style" API, we need to tell InjectLabels(): hands off!
+			//
+			// This upsert tells the ipcache that the prefix is now in the domain of an older user
+			// and thus should not be deleted by clearing createdFromMetadata
+			toUpsert[prefix] = id
+		}
```

Comment:
Outch! Good catch.

But yeah, it should be safe - we basically downgrade the identity from being managed by `InjectLabels` to become manually managed by the caller. Before this PR, all such prefixes were created using the legacy API, so they had `createdFromMetadata = false` from the start. Now, we create the identities via the new API, then call this function here to "downgrade" them to be owned by the old API, and then we call `d.ipcache.RemoveMetadataBatch` for all prefixes that we have downgraded, making them exclusively owned by the old API.

One more indicator that moving the CIDR/FQDN identities to the new IPCache API cannot happen soon enough.

Reply:
Working on it :-)